### PR TITLE
[FLINK-4282]Add Offset Parameter to WindowAssigners

### PR DESCRIPTION
**Pull request by Renkai**
Although there is already a merge request for this issue,I think this implementation is more sensible.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java [hunk 1]
```diff
@@ -52,16 +52,19 @@
 
 	private final long slide;
 
-	protected SlidingEventTimeWindows(long size, long slide) {
+	private final long offset;
+
+	protected SlidingEventTimeWindows(long size, long slide,long offset) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java]: Missing space before `long offset`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java [hunk 2]
```diff
@@ -52,16 +52,19 @@
...
 		this.slide = slide;
+		this.offset = offset;
 	}
 
 	@Override
 	public Collection<TimeWindow> assignWindows(Object element, long timestamp, WindowAssignerContext context) {
 		if (timestamp > Long.MIN_VALUE) {
 			List<TimeWindow> windows = new ArrayList<>((int) (size / slide));
-			long lastStart = timestamp - timestamp % slide;
+			long lastStart = TimeWindow.getWindowStartWithOffset(timestamp,offset,slide);
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java]: Missing spaces after comma.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java [hunk 3]
```diff
@@ -102,7 +105,32 @@ public String toString() {
...
-		return new SlidingEventTimeWindows(size.toMilliseconds(), slide.toMilliseconds());
+		return new SlidingEventTimeWindows(size.toMilliseconds(), slide.toMilliseconds(),0);
+	}
+
+	/**
+	 *  Creates a new {@code SlidingEventTimeWindows} {@link WindowAssigner} that assigns
+	 *  elements to time windows based on the element timestamp and offset.
+	 *<p>
+	 *     For example, if you want window a stream by hour,but window begins at the 15th minutes
+	 *     of each hour, you can use {@code of(Time.hours(1),Time.minutes(15))},then you will get
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java]: Missing space after comma.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 4]
```diff
@@ -553,7 +553,7 @@ public void testSessionWindowsWithCountTrigger() throws Exception {
 				new OneInputStreamOperatorTestHarness<>(operator);
 
 		testHarness.configureForKeyedStream(new TupleKeySelector(), BasicTypeInfo.STRING_TYPE_INFO);
-		
+
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java]: Unrelated whitespace change.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 5]
```diff
@@ -2550,4 +2550,36 @@ public String toString() {
 			return "EventTimeTrigger()";
 		}
 	}
+
+	@Test
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java]: Very good that you added a test for this! 👍 

I'd like to have these in a separate test for `TimeWindow`, however, since the `WindowOperatorTest` file is already very big.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowingTestHarnessTest.java [hunk 6]
```diff
@@ -84,6 +86,84 @@ public void testEventTimeTumblingWindows() throws Exception {
 	}
 
 	@Test
+	public void testEventTimeTumblingWindowsWithOffset() throws Exception {
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowingTestHarnessTest.java]: These tests should go into `WindowOperatorTest`. The existing tests here are just meant to exercise the `WindowTestHarness`.

**Comment by aljoscha** [comment]
The code looks very good! I had some comments about whitespace/formatting and the placement of tests. (It's quite hard to know where to put these without a lot of experience with the code base.)

**Comment by Renkai** [comment]
I have rearranged the code. It is really hard for me to add space after every comma,do you have any code style preset for IDEs so that I can reformat my new code automatically without change those of others?

**Comment by aljoscha** [comment]
@Renkai Unfortunately I'm not aware of such a tool. The IDE I use (IntelliJ) tends to always do changes in a lot of other places when I use the reformat tool so I try to stay away from that.

**Comment by Renkai** [comment]
@aljoscha It's fine.I'll try to write well formatted code manually from now on, and I hope the code in the last commit is well formatted enough.

**Comment by aljoscha** [comment]
I merged it. Thanks for your work! 😃

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/windows/TimeWindow.java [hunk 7]
```diff
@@ -236,6 +236,6 @@ public int compare(TimeWindow o1, TimeWindow o2) {
 	 * @return window start
 	 */
 	public static long getWindowStartWithOffset(long timestamp, long offset, long windowSize) {
```
Comment (soniclavier) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/windows/TimeWindow.java]: Is the windowSize actually slide size? I see that it is called by passing the slide instead of size `TimeWindow.getWindowStartWithOffset(timestamp, offset, slide);`

Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/windows/TimeWindow.java]: In case of sliding windows we have to pass the slide here, yes.
